docs/design/features/byreflike-generics.md

```diff
@@ -9,20 +9,23 @@ Using ByRefLike types in Generic parameters is possible by building upon support
 
 Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
 
+Throws `InvalidProgramException`:
```

Comment:
Should we be more explicit about the allowed timing of these exceptions? I assume that they can either thrown when method is JITed (ie the whole method is replaced with InvalidProgramException) or they can be thrown only once the specific IL instruction is executed. I assume that that the exact timing is unspecified - both options are compliant implementations. Is that correct?

The answer to this question influences where the following is a valid method with predictable behavior or not.
```
string M<T>(T value) where T: allows ref struct
{
     if (typeof(value).IsByRefLike) return "";

     return value.ToString(); // Assume that this is compiled as constrained.callvirt 
}
```

Reply:
> Should we be more explicit about the allowed timing of these exceptions?

I'm not entirely sure in this case. In fact, I'm now struggling to understand what is too much details vs what what is too little. This seems at odds with some of the early feedback about when I was describing what the JIT was going to do (for example, elide instructions). Calling out the JIT or native AOT or "when" this occurs seems to be at odds with the abstraction of IL and getting into when the IL is evaluated or translated. Should I then also consider the behavior in the interpreter case?

Reply:
I think it is important to define the valid IL usages with well-defined behavior and declare everything else as invalid IL with undefined behavior. We do not need to spec the undefined behavior. We like to throw
InvalidProgramException or TypeLoadException for the undefined behavior as a diagnostic aid, but we typically do not guarantee whether the exact exception that is going to be thrown and the undefined behavior can vary between runtimes variants and versions.

It is not clear to me whether these exceptions are part of the well-defined behavior space or undefined behavior space. For example, it is not clear to me whether the following has well-defined behavior:

```
string M<T>(T value) where T: allows ref struct
{
     if (typeof(value).IsByRefLike) return "";

     return value.ToString(); // Assume that this is compiled as constrained.callvirt 
}

M<string>("Hello");
```

Is this a valid program that is guaranteed to work or is this invalid program with undefined behavior?

What about `M<Span<int>>("Hello")`. Is that a valid program that is well-defined behavior or is it an invalid program with undefined behavior?